BASIC NUMPY:
---------------------
(This is a recreation following this: https://sebastianraschka.com/blog/2020/numpy-intro.html)

In [2]:
import numpy as np

In [4]:
print(np.arange(6))

[0 1 2 3 4 5]


In [3]:
def python_dot(x, w):
    z = 0.
    for i in range(len(x)):
        z+= x[i]*w[i]
    return z

In [6]:
array1 = [1., 2., 5.]
array2 = [3., 1., 2.]

In [7]:
print(python_dot(array1, array2))

15.0


Now, let us time a large array:

In [8]:
large1 = list(range(1000))
large2 = list(range(1000))

In [9]:
%timeit python_dot(large1, large2)

59.5 µs ± 1.6 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


* Now, if we use np arrays:*

In [4]:
nplarge1 = np.arange(1000)
nplarge2 = np.arange(1000)

In [5]:
%timeit nplarge1.dot(nplarge2)

1.36 µs ± 45.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


We observe 59.5-->1.36. nparrays are like C arrays, contiguous and homogenous, Helps in cache locality

________________________________________________________________________________________________________

Now, lets look at multidimensional arrays:

In [6]:
pyarray = [1., 2., 3.]

In [7]:
numpyarray = np.array(pyarray)

In [8]:
print(type(numpyarray))

<class 'numpy.ndarray'>


In [9]:
print(numpyarray)

[1. 2. 3.]


In [15]:
pyListOList = [ [1,2,3],
                [4,5,6]]

In [16]:
mat = np.array(pyListOList)

In [17]:
mat = np.array(pyListOList, dtype=int)

In [18]:
type(mat)

numpy.ndarray

In [19]:
print(mat.dtype)

int64


In [21]:
mat_int32 = mat.astype(np.int32)
print(mat_int32.dtype)
type(mat_int32)

int32


numpy.ndarray

Like above, we can convert dtype

In Python, all variables are like references in C++. So, if you want just the value..

In [24]:
print(mat)
mat2 = mat
mat2[0][0] = 5
print(mat)
mat_copy = mat.copy()
mat_copy[0][0] = 4
print(mat)

[[3 2 3]
 [4 5 6]]
[[5 2 3]
 [4 5 6]]
[[5 2 3]
 [4 5 6]]


We have to use ndarry.copy()

nparrays have ndim attribute:

In [25]:
mat.ndim

2

The Shape Attribute: returns nxm as (n, m) {returns a tuple(in case of ndim, so tuple)}

In [26]:
print(mat.shape)

(2, 3)


In [27]:
print(len(mat.shape))

2


In [30]:
assert(len(mat.shape)==mat.ndim)

So, mat.ndim = the length of the shape tuple
_________________________________________________________________________________


NUMPY ARRAY CONSTRUCTION:
-------------------------
Lets look at some array construction routines

1. np.ones(shape, dtype)

In [34]:
ones1 = np.ones((2, 4), dtype=float)
print("Type:{0}".format(type(ones1)))
print("dtype:{0}".format(ones1.dtype))
print("The ndarray:")
print(ones1)

Type:<class 'numpy.ndarray'>
dtype:float64
The ndarray:
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]]


In [36]:
print(np.ones(3))

[1. 1. 1.]


2. np.zeros()

In [37]:
zeros = np.zeros((3,3))
print(zeros)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


In [38]:
print(np.zeros((3,3), dtype=int))

[[0 0 0]
 [0 0 0]
 [0 0 0]]


3. Addition of a scalar to a ndarray:

In [42]:
print(np.zeros((3,3))+23)
print((np.zeros((3,3))+23).astype(int))

[[23. 23. 23.]
 [23. 23. 23.]
 [23. 23. 23.]]
[[23 23 23]
 [23 23 23]
 [23 23 23]]


4. Diagonal matrix:

In [43]:
print(np.diag((1, 2, 3)))

[[1 0 0]
 [0 2 0]
 [0 0 3]]


5. arange fn: (similar to range) gives a half open interval

In [45]:
range1 = np.arange(3, 7, step=2) 
print(range1)
print(np.arange(4, 8)) #4...7
print(np.arange(7))#0...6

[3 5]
[4 5 6 7]
[0 1 2 3 4 5 6]


In [46]:
print(np.arange(2., 5.1, 0.1))

[2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9 3.  3.1 3.2 3.3 3.4 3.5 3.6 3.7
 3.8 3.9 4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9 5. ]


6. linspace fn: Return evenly spaced numbers over a specified interval.

Returns num evenly spaced samples, calculated over the interval [start, stop].

The endpoint of the interval can optionally be excluded.

In [49]:
evenlySpaced = np.linspace(2., 7.)
print(evenlySpaced)
print(np.linspace(2., 7., num=3))

[2.         2.10204082 2.20408163 2.30612245 2.40816327 2.51020408
 2.6122449  2.71428571 2.81632653 2.91836735 3.02040816 3.12244898
 3.2244898  3.32653061 3.42857143 3.53061224 3.63265306 3.73469388
 3.83673469 3.93877551 4.04081633 4.14285714 4.24489796 4.34693878
 4.44897959 4.55102041 4.65306122 4.75510204 4.85714286 4.95918367
 5.06122449 5.16326531 5.26530612 5.36734694 5.46938776 5.57142857
 5.67346939 5.7755102  5.87755102 5.97959184 6.08163265 6.18367347
 6.28571429 6.3877551  6.48979592 6.59183673 6.69387755 6.79591837
 6.89795918 7.        ]
[2.  4.5 7. ]
